pkg/operator/staticpod/controller/guard/guard_controller.go

```diff
@@ -0,0 +1,254 @@
+}
+
+func getGuardPDBName(prefix string) string {
+	return fmt.Sprintf("%s-guard-pdb", prefix)
+}
+
+func (c *GuardController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(5).Info("Syncing guards")
+
+	errs := []error{}
+	shouldDelete := false
+	if c.deleteConditionalFunc != nil && c.deleteConditionalFunc() {
+		shouldDelete = true
+	}
+
+	shouldCreate := false
+	if c.createConditionalFunc != nil && c.createConditionalFunc() {
+		shouldCreate = true
+	}
+
+	switch {
+	case shouldCreate && shouldDelete:
+		errs = append(errs, fmt.Errorf("can not have both create and deletion conditionals evaluated to true at the same time"))
+	case shouldDelete:
+		pdb := resourceread.ReadPodDisruptionBudgetV1OrDie(bindata.MustAsset(filepath.Join("pkg/operator/staticpod/controller/guard", "manifests/pdb.yaml")))
+		pdb.ObjectMeta.Name = getGuardPDBName(c.podResourcePrefix)
+		pdb.ObjectMeta.Namespace = c.targetNamespace
+		_, _, err := resourceapply.DeletePodDisruptionBudget(ctx, c.pdbGetter, syncCtx.Recorder(), pdb)
+		if err != nil {
+			klog.Errorf("Unable to delete PodDisruptionBudget: %v", err)
+			errs = append(errs, err)
+		}
+
+		pods, err := c.podLister.Pods(c.targetNamespace).List(labels.SelectorFromSet(labels.Set{"app": "guard"}))
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			for _, pod := range pods {
+				_, _, err = resourceapply.DeletePod(ctx, c.podGetter, syncCtx.Recorder(), pod)
+				if err != nil {
+					klog.Errorf("Unable to delete Pod: %v", err)
+					errs = append(errs, err)
+				}
+			}
+		}
+	case shouldCreate:
+		selector, err := labels.NewRequirement("node-role.kubernetes.io/master", selection.Equals, []string{""})
+		if err != nil {
+			panic(err)
+		}
+		nodes, err := c.nodeLister.List(labels.NewSelector().Add(*selector))
+		if err != nil {
+			return err
+		}
+
+		pods, err := c.podLister.Pods(c.targetNamespace).List(labels.SelectorFromSet(labels.Set{"app": c.podResourcePrefix}))
+		if err != nil {
+			return err
+		}
+
+		klog.V(5).Infof("Rendering guard pdb")
+
+		pdb := resourceread.ReadPodDisruptionBudgetV1OrDie(bindata.MustAsset(filepath.Join("pkg/operator/staticpod/controller/guard", "manifests/pdb.yaml")))
+		pdb.ObjectMeta.Name = getGuardPDBName(c.podResourcePrefix)
+		pdb.ObjectMeta.Namespace = c.targetNamespace
+		if len(nodes) > 1 {
+			minAvailable := intstr.FromInt(len(nodes) - 1)
+			pdb.Spec.MinAvailable = &minAvailable
+		}
+
+		_, _, err = resourceapply.ApplyPodDisruptionBudget(ctx, c.pdbGetter, syncCtx.Recorder(), pdb)
+		if err != nil {
+			klog.Errorf("Unable to apply PodDisruptionBudget changes: %v", err)
+			return err
+		}
+
+		operands := map[string]*corev1.Pod{}
+		for _, pod := range pods {
+			operands[pod.Spec.NodeName] = pod
+		}
+
+		for _, node := range nodes {
+			if _, exists := operands[node.Name]; !exists {
+				klog.Errorf("Missing operand on node %v", node.Name)
+				errs = append(errs, fmt.Errorf("Missing operand on node %v", node.Name))
```

Comment:
During reboots or similar problems we'll be throwing a lot of errors, I don't think that's the intention. Shouldn't we filter out NotReady nodes earlier in the node query? IIRC installer ignores not ready nodes.

Reply:
Good point. In case the operand pod does not exist, I will skip in case the underlying node is not ready. That might help reduce the number of errors. Also, once a node condition changes, the sync loop gets triggered again so no need to return an error here.